### ISSUE TITLE
TranslatorInterface::getLocale() and setLocale()

### ISSUE THREAD

**Issue by fmonts** [symfony/symfony#29530]
The interface `Symfony\Contracts\Translation`, which replaces `Symfony\Component\Translation` from Symfony 4.2, doesn't have the `getLocale` and `setLocale` methods.

So `$translator->getLocale()` in a controller still works, but throws a warning in the IDE about method not found. 
Since I don't see any reference to this in the UPGRADE file, what should we do?

These methods are useful for example when you are logged in as admin and you're sending an email to an user who has a different language, so the email template needs to be translated into his language, which is different from the current language.

**Comment by nicolas-grekas** [comment]
To get the locale, you should use the request instead.

**Comment by fmonts** [comment]
Also there is a BC break regarding translator: it's not public anymore, `$this->get('translator')` gives:

> Service "translator" not found: even though it exists in the app's container, the container inside "..." is a smaller service locator that only knows about the "doctrine", "form.factory", "http_kernel", "parameter_bag", "request_stack", "router", "security.authorization_checker", "security.csrf.token_manager", "security.token_storage", "session" and "twig" services.Try using dependency injection instead.

**Comment by mdeboer** [comment]
Well yes, to get the requested locale. That is different (from what I understand) from what he's asking. The translation component works independent from the request and should have a getter/setter for the locale, like is implemented but missing from the interface.

**Comment by ro0NL** [comment]
> should have a getter/setter for the locale

IMHO that's an implementation detail of `trans()`, i.e. which _default locale_ to use if `$locale` is not provided. So i think we're striving to follow the principle of least knowledge here.

Framework bundle exposes it's default locale, perhaps you can rely on that?

https://github.com/symfony/symfony/blob/7b03bff57a2782edc6558d57df7ccff99ed0f1a9/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L179

**Comment by ro0NL** [comment]
in that case,  i definitively suggest the alternative solution yes. That avoids messing with the translator state in a global sense.

**Comment by Shoplifter** [comment]
For me it's rather the `getLocale()` that I'm missing. I use it for example in forms to add a locale specific `pattern` attribute for NumberTypes, or to set the locale for a `\Collator` to sort an array or as a parameter for `\IntlDateFormatter::create`. 

As @fmonts stated above, this still works, just the IDE shows an error. You can get rid of this by adding a phpdoc comment something like 
````
public function __construct(TranslatorInterface $translator) {
    /** @var TranslatorTrait $translator **/
    $this->locale = $translator->getLocale();
}
````
Looks strange, but actually the trait is what you get, right?

**Comment by nicolas-grekas** [comment]
@Shoplifter this would be a dirty hack.

Relying on `getLocale()` on the translator means relying on its current mutability.
Since services should be stateless, this should not be the case and that's actually something that we might fix in the future.

Use Request::getLocale() instead.
Closing as there is no bug here - as explained.

**Comment by Shoplifter** [comment]
@nicolas-grekas Thanks for the clarification. Now I know that it had a smell for a reason.

**Comment by dbu** [comment]
as this comes up in searching, i'd like to point out: the use case of changing the locale in all relevant places to e.g. generate an email in the user language or a pdf in the document language, this has been solved in symfony 6.1 with the [locale switcher](https://symfony.com/doc/current/translation.html#switch-locale-programmatically)
